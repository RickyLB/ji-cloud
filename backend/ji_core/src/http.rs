use std::net::{SocketAddr, TcpListener};

/// Creates a [`TcpListener`] via `listenfd` if possible.
///
/// [`TcpListener`]: https://doc.rust-lang.org/stable/std/net/struct.TcpListener.html
pub fn get_tcp_fd() -> Option<TcpListener> {
    #[cfg(feature = "listenfd")]
    let fd = listenfd::ListenFd::from_env().take_tcp_listener(0).unwrap();

    #[cfg(not(feature = "listenfd"))]
    let fd = None;

    fd
}

/// Get the port to run the server on.
pub fn get_addr(default: Option<u16>) -> SocketAddr {
    // 0 is a special port that means "assign me whatever port you want"
    let port = std::env::var("PORT")
        .ok()
        .and_then(|it| it.parse().ok())
        .or(default)
        .unwrap_or(0);

    ([0, 0, 0, 0], port).into()
}
