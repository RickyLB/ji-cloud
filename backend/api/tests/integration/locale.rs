use http::StatusCode;
use macros::test_service;
use sqlx::postgres::{PgConnectOptions, PgPoolOptions};

mod entry;

use crate::{fixture::Fixture, helpers::setup_service};

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_bundles(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/locale/bundle", port))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[test_service(setup = "setup_service", fixtures("Fixture::User", "Fixture::Locale"))]
async fn list_item_kind(port: u16) -> anyhow::Result<()> {
    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/locale/item-kind", port))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body);

    Ok(())
}
