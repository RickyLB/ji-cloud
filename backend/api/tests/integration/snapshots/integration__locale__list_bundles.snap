---
source: tests/integration/locale.rs
expression: body
---
{
  "bundles": [
    {
      "id": "7c11d04e-7c67-11eb-a0d7-9b5fb8f29de9",
      "name": "main page"
    },
    {
      "id": "8046692c-7c67-11eb-a0d7-b7924e4289df",
      "name": "locale"
    },
    {
      "id": "8359a48a-7c67-11eb-a0d7-0fd74777a62c",
      "name": "login"
    },
    {
      "id": "85a46ffe-7c67-11eb-a0d7-277d94fe130c",
      "name": "test 4"
    }
  ]
}
